### ISSUE TITLE
errors in browser console

### ISSUE THREAD

**Issue by beeplin** [feathers-plus/docs.feathers-plus.org#1]
```
client-bundle.js:74 Uncaught TypeError: Cannot set property 'feathersjs' of undefined
    at client-bundle.js:74
    at Object._typeof (client-bundle.js:77)
    at __webpack_require__ (client-bundle.js:20)
    at client-bundle.js:63
    at client-bundle.js:66
(anonymous) @ client-bundle.js:74
_typeof @ client-bundle.js:77
__webpack_require__ @ client-bundle.js:20
(anonymous) @ client-bundle.js:63
(anonymous) @ client-bundle.js:66
guide.html:383 GET http://localhost:4000/images/search.png 404 (Not Found)
```

**Comment by beeplin** [comment]
BTW the `batch-loader` repo is my life saver!! :)

**Comment by beeplin** [comment]
`@feathers-plus/cache` is mentioned in the `batch-loader` docs, but does not actually exist in the `@feathers-plus` scope.

**Comment by eddyystop** [comment]
You can now see the test site at https://feathers-plus.github.io

Alternatively, delete repo/themes/tech-docs/source/js/client-bundle.js and
run hexo generate, hexo server. client-bundle was a test for injecting some
code into the browser.

**Comment by eddyystop** [comment]
I will be copying https://github.com/eddyystop/dataloader-cache-lru to that
name.

**Comment by eddyystop** [comment]
Feel free to start using batch-loader. The code should be solid despite the
lack of tests.

batch-loader and cache were created as dependencies for a new populating
hook which issues roughly an order of magnitude fewer database calls than
`populate`. I can send you the current code if you have a need for it. The
API will likely change but I expect it'll be strongly backwards compatible.

Also planned is a cache hook which will clear batch-loader and cache when a
record mutates.

**Comment by eddyystop** [comment]
batch-loader and cache are just something I'm carrying over to Feathers
from a feathers-graphql service I wrote. That service is fairly
sophisticated as it allows you to make GraphQL queries using FeathersJS
syntax (id, data, params), plus it generates the code (which uses
batch-loader) for the resolvers.

I'll open source that after all the above. I'm not sure you would have a
need for GraphQL.

**Comment by beeplin** [comment]
Having a (at least partly) graphql-compatible syntax is absolutely awesome. That would be much simpler than writing a bunch of awaits or promises (as shown in your final big example in the docs of batcth-loader that populates users, comments and users of comments).

**Comment by beeplin** [comment]
I have being reading the batch-loader docs quite carefully. The vue theme and the whole hexo system works perfectly, except for the errors shown in the browser console. Removing client-bundle just eliminated the first error, and the rest remains.

The theme is so beautiful :)))

**Comment by eddyystop** [comment]
Any suggestions for the API design?

1. Do you find the section `static BatchLoader.getResultsByKey(...)` too busy or hard to read?

2. Arguments which appear in more than one API are not repeated in each API, rather they are listed in the TOC as below. Is that confusing?
```text
Common arguments
  key { String | Number | Object }
  keys { Array<key> }
```

3. Should the Details section only appear when a widget is clicked?

**Comment by eddyystop** [comment]
Are you planning to use batch-loader as a populate hook, or for batch processing?

**Comment by beeplin** [comment]
Let's discuss the API first.

1. `class BatchLoader`'s paramter `options` could have a property `context` but `context` does not show in the property table: 
 
![image](https://user-images.githubusercontent.com/13058150/32544978-2c08b8f0-c4b5-11e7-9cfe-a22d7cf7f472.png)
![image](https://user-images.githubusercontent.com/13058150/32544987-33fd28a2-c4b5-11e7-9c52-081dcc53e178.png)

**Comment by beeplin** [comment]
2. `getResultsByKey` has too many (4 or 5) parameters, which might not be the best design. Maybe we can just pass one object to it as the only paramter: 
```js
getResultsByKey({
  keys: keys,
  data: data,
  identifier: user => user.id, // or some other name
  resultType: '', // or simply `type`
  defaultElement: null,  // optional
  onError: () => {...} // optional
})
```

Furthermore, the `resultTypes` can be more semantic, too. For exmaple, we could just use two options `isCollection` (or `returnCollection`) and `elementRequired`, instead of using a pre-defined `resultType` and then transform it into `collection` and `eleReqd`:

```js
getResultsByKey({
  keys: keys,
  data: data,
  identifier: user => user.id, // or some other name
  returnCollection: true, // or `isCollection`
  elementRequired: false,
  defaultElement: null,  // optional
  onError: () => {...} // optional
})
```

**Comment by beeplin** [comment]
3. I am not sure if it is necessary to have both `load` and `loadMany` at the same time. Personally I prefer just using one `load` to handle both single and multiple record loading. But it is a trivial consideration. :)

**Comment by beeplin** [comment]
4. for the common arguments: 
```
Common arguments
  key { String | Number | Object }
  keys { Array<key> }
```
I do prefer not putting this part at the end of the API docs. Repeating it in multiple parts is surely acceptable and is more reader-friendly.

**Comment by beeplin** [comment]
5. 
> Should the Details section only appear when a widget is clicked?

I prefer all the details sections expanded by default.

**Comment by beeplin** [comment]
6. 
> Do you find the section static BatchLoader.getResultsByKey(...) too busy or hard to read?

Yes, a little bit. And also the beginning part of the `guide`. Once the API design is finalized, I could make a PR later to try to improve these two parts.

**Comment by beeplin** [comment]
7. making `batch-loader` and lru `cache` browser-compatible: Currently `process.nextTick` and `lru-cache` is used, which can only run in `nodejs` environment. We can consider using `setTimeout(fn, 0)` and repo `lru` instead, which can also run in browsers.  (So we can do batching and caching directly from the client side. )

**Comment by eddyystop** [comment]
(1) Thank you. The new populating hook func will use `context` to pass the hook context to the batch-loader func.

(2) I agree with your comments about the number of params. However 4 of those params are required and I would hate to have to type their property names each call.

The resultType options are derived from GraphQL which uses the tags `user`, `user!`, `[user]`, `[user!]`, and `[user!]!`. These resultType options are also used in @feathers-plus/graphql and maybe in the new populating hook. I think the consistency is useful.

(3) In complicated cases you could need to use an object as the key to .load. (Things would get complicated.) So it possible to also imagine a case where an array would be used as the key. So .load and .loadMany keep things clear.

(4) Yes, I am torn. `key` is used in several places. Each time its used we'd need to include
```text
key { String | Number | Object }

The value the batch loader function will use to find the records requested. Its usually a String or Number. An Object may also be used though that would require a more complex batch loade
```
in the Details section. And for some methods we'd have to create a Details section just for this.

**Any further comments?**

(6) I consider batch-loader docs done (except for the cache section). So PR away.

(7) A custom cache needs Map-compatible get, set, delete, clear methods. So you can use `lru` on the client.

nextTick and setTimeout behave differently. There is no guarantee when 'suspended' promise chains would 'restart' with setTimeout. **Can you think of a solution other than having both a Node and browser wrapper injecting nextTick/setTimeout to a core module?** And that would unfortunately require a webpack build step.

**Comment by eddyystop** [comment]
Thanks for your worthwhile comments on batch-loader. I however wasn't clear in my request. I was asking for your comments on the design of the API docs themselves. That needs to be settled before we start documenting other modules, especially the common hooks.

Any followup to (4) above? Any other comments?

**Comment by beeplin** [comment]
For the design of the API docs (and also for the (4)), I do think each API should have its own `Arguments` part, `Usage` part (and `Returns` and `Example` part if applicable). Common Arguments should be moved back to their corresponding positions, although it seems a little bit redundant. A full API should be like this: 
![image](https://user-images.githubusercontent.com/13058150/32559186-7fead6c0-c4e1-11e7-8ae7-da98f44ab575.png)
I could take care of this in my PR tomorrow.

**Comment by beeplin** [comment]
For the `process.nextTick`/`setTimeout` issue, since we only have `setTimeout` in browser, I guess we could just detect the current environment first, and then choose which one to use. Something like this: 

```js
const nextTick = process && (typeof process.nextTick === 'function') 
  ? process.nextTick 
  : fn => setTimeout(fn, 0);

...

resolvedPromise.then(() => nextTick(fn));
```
(note that `process` is a pre-defined global object and doesn't need to be required.)

I am not familiar with the difference between `process.nextTick` and `setTimeout`, so not sure if this is what you want.

**Comment by eddyystop** [comment]
vm.$watch is a much better Vue template than any I had found. Thanks!

I had forgotten process was a global. Thanks also.

**Comment by eddyystop** [comment]
vm.$watch is a much better template than any I had found. Thanks!

I had forgotten process was a global. Thanks!

**Comment by eddyystop** [comment]
vm.$watch is a much better template than any I had found. Thanks!

I had forgotten process was a global. Thanks!

**Comment by eddyystop** [comment]
vm.$watch is a much better template than any I had found. Thanks!

I had forgotten process was a global. Thanks!

**Comment by beeplin** [comment]
@eddyystop please check the PR #3. :)

I am eager to see your works for using GraphQL and `batch-loader` to do batching and population. :) I would like to use it in my current project if it is coming soon.

**Comment by beeplin** [comment]
I am also eager to see a new population hook using batch-loader. For those who are not familiar with graphql, the population hook might be an easier choice.

**Comment by eddyystop** [comment]
I sent you a GitHub invite to the currently private feathers-graphql repo. It'll be a longish read.

The populate hooks give the server control of population. GraphQL is a very client oriented approach. I think GraphQL may be most useful when you have separate front and backend teams. The frontend developers may not use good practices getting their data from the backend DB, even using FeathersJS.

I suspect the frontend widgets should be using GraphQL `fragments` to specify their needs, allowing GraphQL to coalesce them into 1 request. I am yet to do a deep-dive into fragments, though eddyystop/graphql-resolvers-ast has allowed me to start exploring the details.

GraphQL will always make 1 roundtrip with all the data the frontend then needs. I found your desire to use batch-loader on the client interesting. If the client caches enough information to handle most requests without calling the backend (and what about mutation?), then a client batch-loader would be more efficient and faster. If a batch-loader design still calls the backend even once on most requests, then GraphQL may be the way to go. GraphQL will send all the data for the entire request, so the payload may be bigger than, say, using feathers-batch.

Code that's ammenable to using batch-loader, may need significant refactoring to use GraphQL.

I agree that GraphQL may be a lot for many people to absorb. The API of the new populating hook keeps changing to find a balance between GraphQL ideas and ease of concept and use.

**Comment by eddyystop** [comment]
I sent an invite to the currently private new populating hook. **I'm interested in your reaction to the high-level design of the schema.** The current populate hook's includes eventually come down to executing a .find() call. Initially the parentField and childField schema fields were used to form the params. Then reasonable use cases required a more complex .find() So provider, innerPopulate, query, and select schema fields were introduced. Then developers had to wrap their heads around these schema fields.

This hook's design is basically to allow the dev to provide the .find() needed in pure FeathersJS (within the batch-loader func), so the hook does not need to build or modify it. Also no explaination of options is needed for the .find() as it's pure FeathersJS.

I'm thinking of expanding to a (schema, request) signature. Request would specify what subset of the schema definition to return. The default would be everything. This would allow the client to dynamically specify what's needed. Or the server can do it based on client permissions.

I think changing schema to be more like GraphQL types, e.g. [users!]!, would be going too far. First, it being a hook means there is no Query type, the data in context.result is the result of what have been the Query. Second, you might as well use feathers-query if you want to go this far.

p.s. Look at tests/basic-await.test.js to get up to speed quickly.

**Comment by eddyystop** [comment]
I have changed the batch-loader API docs to more closely match Vue.$watch. Let me know what you think of the new layout vs the `Details` section we used before. Those `Detail` section are still in the API docs, and will be removed if you like the newer look.

**Comment by beeplin** [comment]
For the API docs I prefer the previous layout with `Details`. Using tables to list the types, default values and explanations of arguments/options are easier to read. 

I will have a look at the new super-population ASAP. I am definitely planning to use batch-loader at the server side to do population. As for the client side and GraphQL... that's a long story with lots of hard decisions to make... Will discuss that later, maybe tomorrow. It is late night in China now.. :)

**Comment by beeplin** [comment]
I roughly checked the source code and the tests of `super-populate-dev`. Maybe because I am too used to the syntax of the old `populate` hook, I feel that the new API is harder for ordinary developers to handle.

The main reason is that, for most people, *declarative* programming is always easier and safer (fewer bugs) then *imperative* programming. The new imperative (pure JS logic with explicit and heavy use of batch-loader), of coz, is much more flexible -- that's a trade-off between simplicity and flexibility. On the other side, the old API uses a logic with `nameAs`, `parentField`, `childField`, and similar logics are already very common in SQL joins and mongdb joins and etc. 

I do like the new API with `before/after/include` schema which is very overwhelmingly powerful. While at the same time, personally I would also like to be able to use a declarative  API just like the old one, with batch-loader running behind it to improve its performance. (To achieve that, maybe the `select` function and the `query` object can only have properties beginning with $ (like $limit, $sort, $select, etc.). ) I am not sure if it is possible that the new imperative API and the old declarative API could coexist and people can choose which one to use according to their specific needs.

**Comment by eddyystop** [comment]
I initially felt the same way about moving from declarative/DSL to imperative. But is this (from the docs) better than coding a .find()?
``` js
{ service: 'comments',
  nameAs: 'commentItems',
  permissions: '...',
  parentField: 'id',
  childField: 'postId',
  query: {
    $limit: 5,
    $select: ['title', 'content', 'postId'],
    $sort: {createdAt: -1}
  },
  select: (hook, parent, depth) => ({ $limit: 6 }),
  asArray: true,
  paginate: false,
  provider: undefined,
  useInnerPopulate: false,
  include: [ ... ]
}
```

We still can't select which fields in the record to return, possibly increasing the number of TCP/IP packets.

We still can't customize context.params beyound query.
- Can't pass thru adapter specific params to Mongoose, etc.
- Can't control the authentication and permission attributes. (The permissions prop is unused, another issue.)

We can always introduce some props for the above, but more use cases could come up.

**Comment by eddyystop** [comment]
The hook now supports the additional concept of a request (object or thunk, default whole schema). So you can do
``` js
const includeComputed = true;
const includeUsers = true;
const includeComments = true;

superPopulate(schema, context => ({
  compute1: includeComputed,
  userRecord: true,
  commentRecords: includeComments && { comment_userRecord: includeUsers },
  starIdsRecords: includeUsers,
  reputation_UserIdRecord: includeUsers,
}))
```

Computed fields can be defined in the includes, e.g.
``` js
{
  name: 'compute1',
  populateParent: (post, _, context) => post.computedId = `id: ${post.id}` // sync now supported
}
```

Both concepts are GraphQL-ish. See tests/design-1.js. A corollary effect is that batch-loaders are now optional. You can directly use FeathersJS method calls though, of course, performance will suffer.
``` js
populateParent: async (post, _, context) =>
  post.userRecord = await users.find({ query: { id: post.userId } }),
```

**Comment by eddyystop** [comment]
I thought about putting a declarative wrapper around at least
``` js
user: new BatchLoader(async (keys, context) => {
    const result =  users.find(makeCallingParams(context, { id: { $in: getUniqueKeys(keys) } }))
    return getResultsByKey(keys, result, user => user.id, '!');
  },
  { context }
)
```

My sense is it would be a level of abstraction that's over engineered. I'm also concerned many people would try it which I think is not best for either them or my support load.

**Comment by eddyystop** [comment]
The `request` object can now include arguments to pass to the resolver functions. Its still optional.
``` js
const includeComputed = true;
const includeUsers = true;
const includeComments = true;

const request = {
  compute1: includeComputed && [1, 2],
  userRecord: includeUsers,
  commentRecords: includeComments && { args: [3, 4], comment_userRecord: includeUsers && [5, 6] },
  starIdsRecords: includeUsers && [7, 8],
  reputation_UserIdRecord: includeUsers && [9, 10, 11],
};
```

This hook is now threatening  to be a poor man's GraphQL. Perhaps we even ought to call it `graphql`. LOL. I can't think of other GraphQL features I'm itching to include, but perhaps you can think of some.

Frankly there appears to be huge power here, and just look at the LOC in the hook!

**Comment by beeplin** [comment]
My suggestion is to have both `populate` and `super-populate` coexisting, `populate` (powered by `batch-loader`) to cover 90% of common use cases, and `super-populate` to handle some rare and complicated use cases. 

I have a feeling that you tend to make the `super-populate` hook more and more omnipotent -- I saw in the new `super-populate` hook that we can even add computed fields in it (in the old hooks we do this by writing our own small hook functions or using the `serialize` hook following the `populate` hook). You also mentioned:

> We still can't select which fields in the record to return, possibly increasing the number of TCP/IP packets.

But I think that's not the job of `super-populate` -- a `pluck` or `keep` hook following it could do this job very well.

What I am interested in is a hook whose major job is to *load and merge data from other services* -- perhaps it is better to be called `join` rather than `populate`. The old `populate` hook is exactly such a hook.

And I don't think the syntax of the old `populate` hook is too complicated. In most cases we don't need to use all the options of the hook. For me `select`, `permissions` are very rarely used (I use `feathers-permisions` to handle permissions --  again, just use `populate` for data population/join.) A typical usage is just like this: 

```js
const schema = {
  service: 'posts',
  include: [
    {
      service: 'users',
      nameAs: 'author',
      parentField: 'authorId',
      childField: 'id',
      query: {
        $select: ['name']
      },
    },
    {
      service: 'users',
      nameAs: 'starers',
      parentField: 'starerIds',
      childField: 'id',
      query: {
        $select: ['name']
      },
      asArray: true,
    },
    {
      service: 'comments',
      parentField: 'id',
      childField: 'postId',
      query: {
        $limit: 5,
        $select: ['title', 'content', 'postId'],
        $sort: {createdAt: -1}
      },
      asArray: true,
      include: [{
        service: 'users',
        nameAs: 'author',
        parentField: 'authorId',
        childField: 'id',
        query: {
          $select: ['name']
        }
      }]
    },
  ],
}
```
In my option the DSL above is quite straightforward and its semantics is quite clear.

The only problem of the old `populate` hook is its performance, and `batch-loader` is exactly the right tool to improve that. 

Therefore, I am glad to see the new `super-populate` hook which is actually a wrapper of user-customized logics and can handle very complicated cases, and on the other hand I am more interest in a new `populate` hook which keeps the current API but is powered by `batch-loader` and is no longer suffering from performance.

**Comment by eddyystop** [comment]
Thank you for your thoughts. I always consider them worth considering.

I don't see myself putting more work into the current `populate` because, first, I don't have the time and, second, I feel its a dead-end conceptually.

I would be happy to accept a PR should you decide to implement batch-loader in `populate`.

**Comment by eddyystop** [comment]
Do you have any suggestions for naming this new hook?
- super-populate may be a bit much.
- fast-populate doesn't imply all its capabilities
- graphql seems presumputous and isn't accurate

**Comment by eddyystop** [comment]
I made the changes you suggested on the API docs. https://feathers-plus.github.io/v1/batch-loader/ would be the template we follow for all API docs, if you have no more suggestions.

Thanks for you help, its noticeably improved the docs.

**Comment by beeplin** [comment]
No further suggestions for the API docs. :)

I will try to make a PR for the old populate hook.

**Comment by eddyystop** [comment]
I've thought about your comments.

This would correspond to your non-batch-loader `populate` example, though I've highlighted more flexibility.
``` js
const commentSchema = {
  joins: {
    author: $select => async comment => comment.author = (await users.find({ query: { 
      id: comment.authorId, $select: $select || ['name']
    } }))[0],
} };

const postSchema = {
  joins: {
    author: $select => async post => post.author = (await users.find({ query: {
      id: post.authorId, $select: $select || ['name']
    } }))[0],
    starers: $select => async post => post.starers = await users.find({ query: {
      id: { $in: post.starerIds }, $select: $select || ['name']
    } }),
    comments: {
      resolver: ($select, $limit, $sort) => async post => post.comments = await comments.find({ query: {
        postId: post.id, $select: $select || ['title', 'content', 'postId'], $limit: $limit || 5, [$sort]: { createdAt: -1 }
      } }),
      joins: commentSchema
} } };
```

I think you would use similar ideas to introduce **shared** batch-loaders -- and you do need to share them. So no reduced complexity there.

You'll have to adapt this imperative design should you want to use GraphQL.

**Comment by eddyystop** [comment]
@beeplin, I've written a draft of the guide for `fastJoin`. Its not edited at all yet. If you should have the time, I would appreciate your thoughts on the outline.
